File: Sem2/PO_project/python/hex_board_panel.py
from __future__ import annotations

import math
from typing import Callable, Optional

import tkinter as tk


class HexBoardPanel(tk.Canvas):
    def __init__(self, master, game, size: int = 16, **kwargs) -> None:
        self.game = game
        self.size = size
        self.on_cell_click: Optional[Callable[[int, int], None]] = None
        self.hexes: dict[tuple[int, int], list[tuple[float, float]]] = {}
        self.padding_x = size
        self.padding_y = size
        self._build_hexes()
        super().__init__(master, background="black", highlightthickness=0, **kwargs)
        self.configure(width=self.width, height=self.height)
        self.bind("<Button-1>", self._handle_click)
        self.redraw()

    def set_on_cell_click(self, handler: Callable[[int, int], None]) -> None:
        self.on_cell_click = handler

    def _build_hex(self, cx: float, cy: float, radius: float) -> list[tuple[float, float]]:
        points: list[tuple[float, float]] = []
        for i in range(6):
            angle = math.radians(60 * i - 30)
            points.append((cx + radius * math.cos(angle), cy + radius * math.sin(angle)))
        return points

    def _build_hexes(self) -> None:
        self.hexes.clear()
        rows = self.game.get_board().rows
        cols = self.game.get_board().cols

        width_step = math.sqrt(3) * self.size
        height_step = 1.5 * self.size
        hex_height = 2 * self.size

        for r in range(rows):
            for c in range(cols):
                cx = self.padding_x + width_step * (c + 0.5 * (r & 1))
                cy = self.padding_y + height_step * r
                self.hexes[(r, c)] = self._build_hex(cx, cy, self.size)

        self.width = int(self.padding_x * 2 + width_step * (cols + 0.5))
        self.height = int(self.padding_y * 2 + height_step * (rows - 1) + hex_height)

    def _color_to_hex(self, color: tuple[int, int, int]) -> str:
        return "#%02x%02x%02x" % color

    def _find_cell(self, x: float, y: float) -> Optional[tuple[int, int]]:
        for cell, poly in self.hexes.items():
            if self._point_in_polygon(x, y, poly):
                return cell
        return None

    @staticmethod
    def _point_in_polygon(x: float, y: float, polygon: list[tuple[float, float]]) -> bool:
        inside = False
        count = len(polygon)
        j = count - 1
        for i in range(count):
            xi, yi = polygon[i]
            xj, yj = polygon[j]
            intersects = ((yi > y) != (yj > y)) and (
                x < (xj - xi) * (y - yi) / ((yj - yi) or 1e-9) + xi
            )
            if intersects:
                inside = not inside
            j = i
        return inside

    def _handle_click(self, event) -> None:
        cell = self._find_cell(event.x, event.y)
        if cell is not None and self.on_cell_click is not None:
            self.on_cell_click(cell[0], cell[1])

    def redraw(self) -> None:
        self.delete("all")
        for (r, c), poly in self.hexes.items():
            org = self.game.get_organism_at(r, c)
            fill = self._color_to_hex(org.get_color()) if org is not None else "#000000"
            points = [coord for point in poly for coord in point]
            self.create_polygon(*points, fill=fill, outline="#282828", tags=("cell", f"cell:{r}:{c}"))

File: Sem2/PO_project/Java/HexBoardPanel.java
package Java;

import javax.swing.*;
import java.awt.*;
import java.awt.event.MouseAdapter;
import java.awt.event.MouseEvent;
import java.util.HashMap;
import java.util.Map;
import java.util.function.BiConsumer;

public class HexBoardPanel extends JPanel {
    private final Game game;
    private final int size;
    private final Map<Point, Polygon> hexes = new HashMap<>();
    private BiConsumer<Integer, Integer> onCellClick;

    public HexBoardPanel(Game game, int size) {
        this.game = game;
        this.size = size;
        buildHexes();
        setBackground(Color.black);

        addMouseListener(new MouseAdapter() {
            @Override
            public void mouseClicked(MouseEvent e) {
                Point cell = findCell(e.getPoint());
                if (cell != null && onCellClick != null) {
                    onCellClick.accept(cell.x, cell.y);
                }
            }
        });
    }

    public void setOnCellClick(BiConsumer<Integer, Integer> handler) {
        this.onCellClick = handler;
    }

    private Polygon buildHex(double cx, double cy, double r) {
        int[] xs = new int[6];
        int[] ys = new int[6];
        for (int i = 0; i < 6; i++) {
            double angle = Math.toRadians(60 * i - 30);
            xs[i] = (int)Math.round(cx + r * Math.cos(angle));
            ys[i] = (int)Math.round(cy + r * Math.sin(angle));
        }
        return new Polygon(xs, ys, 6);
    }

    private void buildHexes() {
        hexes.clear();
        int rows = game.getBoard().getRows();
        int cols = game.getBoard().getCols();

        double w = Math.sqrt(3) * size;
        double h = 2 * size;
        double vStep = 1.5 * size;

        double padX = size;
        double padY = size;

        for (int r = 0; r < rows; r++) {
            for (int c = 0; c < cols; c++) {
                double cx = padX + w * (c + 0.5 * (r & 1));
                double cy = padY + vStep * r;
                Polygon p = buildHex(cx, cy, size);
                hexes.put(new Point(r, c), p);
            }
        }

        int width = (int)(padX * 2 + w * (cols + 0.5));
        int height = (int)(padY * 2 + vStep * (rows - 1) + h);
        setPreferredSize(new Dimension(width, height));
    }

    private Point findCell(Point p) {
        for (Map.Entry<Point, Polygon> e : hexes.entrySet()) {
            if (e.getValue().contains(p)) {
                return e.getKey();
            }
        }
        return null;
    }

    @Override
    protected void paintComponent(Graphics g) {
        super.paintComponent(g);

        for (Map.Entry<Point, Polygon> e : hexes.entrySet()) {
            Point cell = e.getKey();
            Polygon poly = e.getValue();
            Organism org = game.getOrganismAt(cell.x, cell.y);

            if (org != null) {
                g.setColor(org.getColor());
                g.fillPolygon(poly);
            } else {
                g.setColor(Color.black);
                g.fillPolygon(poly);
            }

            g.setColor(new Color(40, 40, 40));
            g.drawPolygon(poly);
        }
    }

}
